package winkle.splendor;

message Room {
  optional strign id = 1;
  optional Game game = 2;
  optional uint64 create_time = 3;
  repeated Audience audience = 4;
}

message Game {
 optional int64 id = 1;
 enum State {
  STATE_UNKNOWN = 0;
  STATE_READY_START = 1;
  STATE_PLAYING = 2;
  STATE_LAST_ROUND = 3;
  STATE_END = 4;
 };
 optional State state = 2;
 optional int32 player_turn = 3;
 optional Desk desk = 4;
 repeated Player players = 5;
 optional int64 last_update_ts = 6;
}

message Audience {
  optional uint64 id = 1; 
}

message PlayerAction {
  enum Order {
    NOTHING = 0;
    GET_GEMS = 1;
    BUY_CARD = 2;
    BUY_NOBLE = 3;
  };
}

message Desk {
 repeated string level = 1;
 repeated Card cards = 2;
 repeated Noble nobles = 3;
 repeated Gems gems = 4;
}

message Player {
  optional int64 id = 1;
  optional string user_name = 2;
  optional int32 match_index = 3 [default = 1];
  optional bool is_ready = 4 [default = false];
  optional bool is_active = 5 [default = false];
}

message Card {
  optional int64 id = 1;
  optional Gems value_gems = 2;
  optional Gems cost_gems = 3;
  optional int32 score = 4;
}

message Gems {
  enum Color {
    RUBY = 1;  // 红
    EMERALD = 2; // 绿
    SAPPHIRE = 3; // 蓝
    DIAMOND = 4;
    CHOCOLATE = 5; //黑
    GOLD = 6;
  };
  optional Color color = 1;
}

message Noble {
  optional int32 id = 1;
  repeated Gems cost_gems = 2;
  optional int32 score = 3;
}